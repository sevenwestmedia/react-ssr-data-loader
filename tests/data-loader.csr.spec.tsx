import Adapter from 'enzyme-adapter-react-16'
import { ComponentFixture } from './helpers/component-fixture'
import { configure } from 'enzyme'
import { processEventLoop } from './helpers/event-loop-helpers'
import { act } from 'react-dom/test-utils'

configure({ adapter: new Adapter() })

describe('Client side render', () => {
    it('should start loading data if not loaded', () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })

        sut.assertState()
    })

    it('should pass loaded data once promise resolves', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })

        await act(() =>
            sut.testDataPromise.resolve({
                result: 'Success resolved!',
            }),
        )

        sut.assertState()
    })

    it('loads data when props change', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })

        await act(async () => {
            await sut.testDataPromise.resolve({ result: 'Success props change!' })
            sut.resetPromise()
            sut.setId('newData')
            await processEventLoop()
            await sut.testDataPromise.resolve({ result: 'Success props change 2!' })
        })

        sut.assertState()
    })

    it('should pass failure when data load fails', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })

        await act(() => sut.testDataPromise.reject(new Error('Boom!')))

        const lastAction = sut.getState().renderProps.lastAction
        expect(lastAction.success).toBe(false)
        expect(lastAction.type).toBe('fetch')
        if (!lastAction.success) {
            expect(() => {
                throw lastAction.error
            }).toThrowError('Boom!')
        }
    })

    it('client render after SSR with data should not fetch data', async () => {
        let sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: true })

        await act(() => sut.testDataPromise.resolve({ result: 'Success from server!' }))

        sut = new ComponentFixture(sut.currentState, 'testKey', { isServerSideRender: false })

        sut.assertState()
    })

    it('should remove data from redux when unmounted', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })

        await act(async () => {
            await sut.testDataPromise.resolve({ result: 'Success removed!' })
            await sut.unmount()
        })

        const lastEvent = sut.events[sut.events.length - 1]
        expect(lastEvent).toEqual({ state: {}, type: 'state-changed' })
    })

    it('should ignore completion once component is unmounted', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })

        await act(async () => {
            await sut.unmount()
            await sut.testDataPromise.resolve({ result: 'Success unmounted!' })
        })

        expect(sut.events).toEqual([
            { state: {}, type: 'state-changed' },
            expect.objectContaining({ type: 'state-changed' }),
            expect.objectContaining({ type: 'begin-loading-event' }),
            { state: {}, type: 'state-changed' },
            expect.objectContaining({ type: 'end-loading-event' }),
            expect.objectContaining({ type: 'data-load-completed' }),
        ])
    })
})
