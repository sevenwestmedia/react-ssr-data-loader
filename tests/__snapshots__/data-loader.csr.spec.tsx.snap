// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client side render client render after SSR with data should not fetch data 1`] = `
Object {
  "events": Array [],
  "loadAllCompletedCalled": 0,
  "loadDataCount": 0,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 1,
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": true,
      "hasData": true,
      "result": Object {
        "result": "Success from server!",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`Client side render loads data when props change 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Success props change!",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "44f8a95a": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "44f8a95a",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "44f8a95a": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Success props change 2!",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "44f8a95a",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "44f8a95a",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
  ],
  "loadAllCompletedCalled": 2,
  "loadDataCount": 2,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 4,
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Success props change 2!",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`Client side render should pass loaded data once promise resolves 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Success resolved!",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
  ],
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Success resolved!",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`Client side render should start loading data if not loaded 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
  ],
  "loadAllCompletedCalled": 0,
  "loadDataCount": 1,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 1,
  "renderProps": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
}
`;
